registry/registry.go

```diff
@@ -16,28 +17,68 @@ import (
 	"github.com/remind101/empire/procfile"
 )
 
-// PullAndExtract returns a ProcfileExtractor that will pull the image using the
-// docker client, then attempt to extract the the Procfile, or fallback to the
-// CMD directive in the Dockerfile.
-func PullAndExtract(c *dockerutil.Client) empire.ProcfileExtractor {
+// DigestError can be returned if the Docker image doesn't have an RepoDigests.
+type DigestError struct {
+	Image string
+}
+
+// Error implements the error interface.
+func (e *DigestError) Error() string {
+	return fmt.Sprintf("image %s has no digests", e.Image)
+}
+
+// dockerDaemon is an implementation of the empire.ImageRegistry interface
+// backed by a local Docker daemon.
+type dockerDaemon struct {
+	docker *dockerutil.Client
+}
+
+// DockerDaemon returns an empire.ImageRegistry that uses a local Docker Daemon
+// to extract procfiles and resolve images.
+func DockerDaemon(c *dockerutil.Client) empire.ImageRegistry {
+	return &dockerDaemon{c}
+}
+
+func (r *dockerDaemon) ExtractProcfile(ctx context.Context, img image.Image, w io.Writer) ([]byte, error) {
+	if err := r.docker.PullImage(ctx, docker.PullImageOptions{
+		Registry:      img.Registry,
+		Repository:    img.Repository,
+		Tag:           img.Tag,
+		OutputStream:  w,
+		RawJSONStream: true,
+	}); err != nil {
+		return nil, err
+	}
+
 	e := multiExtractor(
-		newFileExtractor(c),
-		newCMDExtractor(c),
+		newFileExtractor(r.docker),
+		newCMDExtractor(r.docker),
 	)
+	return e.ExtractProcfile(ctx, img, w)
+}
 
-	return empire.ProcfileExtractorFunc(func(ctx context.Context, img image.Image, w io.Writer) ([]byte, error) {
-		if err := c.PullImage(ctx, docker.PullImageOptions{
-			Registry:      img.Registry,
-			Repository:    img.Repository,
-			Tag:           img.Tag,
-			OutputStream:  w,
-			RawJSONStream: true,
-		}); err != nil {
-			return nil, err
-		}
+// Resolve returns an immutable identifier for the given image. It does so by
+// inspecting the image using the Docker Daemon API, and returns the first repo
+// digest it finds.
+func (r *dockerDaemon) Resolve(ctx context.Context, img image.Image) (image.Image, error) {
+	// If the image already references an immutable identifier, there's
+	// nothing for us to do.
+	if img.Digest != "" {
+		return img, nil
+	}
 
-		return e.Extract(ctx, img, w)
-	})
+	i, err := r.docker.InspectImage(img.String())
+	if err != nil {
+		return img, err
+	}
+
+	if len(i.RepoDigests) <= 0 {
+		return img, &DigestError{Image: img.String()}
```

Comment:
Turns out RepoDigests is only populated from a `docker pull` in Docker 1.12+. This thread https://github.com/moby/moby/issues/15508 explains the details of that.

We've been locked to 1.11.x for some time now, just for stability, but should probably consider upgrading to a later version.

Instead of returning an error here, this could just fallback to the existing behavior instead.

Reply:
I was just wondering about this - yeah, seems like falling back to existing behavior - or possibly having it be able to configure that you require digests - would be good.